### PR TITLE
Add CBMC harness for QueueGenericCreate

### PR DESCRIPTION
**Pull request by mmuesly**
Description
-----------

Checklist:
----------

- [ ] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

### CODE REVIEW

**Comment by RichardBarry** [comment]
Fantastic we are moving onto the kernel with these proofs now, however there are a few issues with these changes:

- Please don't introduce a new header file as the code builds with some 20 odd different compilers and we can't guarantee the header file is going to be there (stdint.h is not provided with all of them).  In this case I think using portMAX_DELAY in place of INT_MAX will work.

Related to:

configASSERT( uxItemSize == 0
			|| (uxQueueLength > 0 && uxItemSize >= 0 
				&& uxItemSize < (INT_MAX - sizeof(Queue_t))/uxQueueLength) );

- 0 is the value used each time a semaphore is created as semaphores have length (counting semaphores for example) but no actual storage, hence 0 is a valid value.  

- Please use a separate assert for each set of values being checked, rather than combing them in one.  This is per the coding standard (maybe not documented) and is something folks were asking about on the forum just last week.

- The statement is quite complex so please add a comment to say the condition it is checking for.  In this case I think the check to ensure the value is not above INT_MAX is probably obsolete in practice (nothing has that much memory in the MCU space), although may be required to claim formal proof.

**Comment by mmuesly** [comment]
Thank you for the comment.

I think the portMAX_DELAY is a good suggestion instead of INT_MAX and does the job without further header dependencies.

Further, I minimized the assert statement to a single line and added a comment as requested. 

The new assert statement ensures that all parameters are set to a range that allow secure algebraic execution of the method without under or overflow errors. Therefore it must hold for a formal analysis.

**Comment by mmuesly** [comment]
This PR depends on: #697 and #774

File: tools/cbmc/include/cbmc.h [hunk 1]
```diff
@@ -25,3 +25,26 @@ enum CBMC_LOOP_CONDITION { CBMC_LOOP_BREAK, CBMC_LOOP_CONTINUE, CBMC_LOOP_RETURN
 
 #define __CPROVER_printf_ptr(var) { uint8_t *ValueOf_ ## var = (uint8_t *) var; }
 #define __CPROVER_printf2_ptr(str,exp) { uint8_t *ValueOf_ ## str = (uint8_t *) (exp); }
+
+/* CBMC assert to test pvPortMalloc result when xWantedSize is 0. Mostly used to report
```
Comment (danielsn) [inline tools/cbmc/include/cbmc.h]: Is this a guarantee of the FreeRTOS system? If so, make a comment referencing where this is documented to be true? Its not true for regular malloc.

File: tools/cbmc/include/cbmc.h [hunk 2]
```diff
@@ -25,3 +25,26 @@ enum CBMC_LOOP_CONDITION { CBMC_LOOP_BREAK, CBMC_LOOP_CONTINUE, CBMC_LOOP_RETURN
...
+ */
+#define __CPROVER_assert_zero_allocation() __CPROVER_assert( pvPortMalloc(0) == NULL, "pvPortMalloc allows zero-allocated memory.")
+
+/* xWantedSize is not bounded in this function, but there might be a need to bound it in the future.
+ * In theory, CBMC malloc allows to allocate an arbitrary amount of data. This will not be true for
+ * embedded devices.
+ */
+void *pvPortMalloc( size_t xWantedSize )
+{
+	if ( xWantedSize == 0 )
```
Comment (danielsn) [inline tools/cbmc/include/cbmc.h]: See above.  This is not true of malloc in general.

File: tools/cbmc/proofs/Queue/QueueGenericCreate/Configurations.json [hunk 3]
```diff
@@ -0,0 +1,68 @@
+{
+  "ENTRY": "QueueGenericCreate",
+  "CBMCFLAGS":
+  [
+      "--unwind 1",
+      "--signed-overflow-check",
+      "--pointer-overflow-check",
+      "--unsigned-overflow-check"
```
Comment (danielsn) [inline tools/cbmc/proofs/Queue/QueueGenericCreate/Configurations.json]: no memory safety flags?

File: tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+#include "queue.h"
+#include "queue_datastructure.h"
+#include "cbmc.h"
+
+
+void harness(){
+	UBaseType_t uxQueueLength;
+	UBaseType_t uxItemSize;
+	uint8_t ucQueueType;
+	__CPROVER_assume(uxQueueLength > ( UBaseType_t ) 0);
```
Comment (danielsn) [inline tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c]: What happens if the user tried to create a queue of length 0?  Would that be legal?

File: tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+	/* The implicit assumption for the QueueGenericCreate method is,
+	   that there are no overflows in the computation and the inputs are safe.
+	   There is no check for this in the code base */
```
Comment (danielsn) [inline tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c]: Then add some!

File: tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+	UBaseType_t upper_bound = portMAX_DELAY - sizeof(Queue_t);
```
Comment (danielsn) [inline tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c]: Why this upper bound? Is this for performance reasons, or required by the spec?

File: tools/cbmc/proofs/make_type_header_files.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python3
```
Comment (danielsn) [inline tools/cbmc/proofs/make_type_header_files.py]: This does not belong in this PR.  refactor into a seperate PR

File: tools/cbmc/include/cbmc.h [hunk 8]
```diff
@@ -25,3 +25,26 @@ enum CBMC_LOOP_CONDITION { CBMC_LOOP_BREAK, CBMC_LOOP_CONTINUE, CBMC_LOOP_RETURN
 
 #define __CPROVER_printf_ptr(var) { uint8_t *ValueOf_ ## var = (uint8_t *) var; }
 #define __CPROVER_printf2_ptr(str,exp) { uint8_t *ValueOf_ ## str = (uint8_t *) (exp); }
+
+/* CBMC assert to test pvPortMalloc result when xWantedSize is 0. Mostly used to report
+ * full coverage on pvPortMalloc, but use with caution as it might complicate debugging
+ */
+#define __CPROVER_assert_zero_allocation() __CPROVER_assert( pvPortMalloc(0) == NULL, "pvPortMalloc allows zero-allocated memory.")
```
[the remark this answers is not in the document]
Comment (mmuesly) [reply tools/cbmc/include/cbmc.h]: I think, this discussion should be part of #774.

File: tools/cbmc/include/cbmc.h [hunk 9]
```diff
@@ -25,3 +25,26 @@ enum CBMC_LOOP_CONDITION { CBMC_LOOP_BREAK, CBMC_LOOP_CONTINUE, CBMC_LOOP_RETURN
...
+
+/* xWantedSize is not bounded in this function, but there might be a need to bound it in the future.
+ * In theory, CBMC malloc allows to allocate an arbitrary amount of data. This will not be true for
+ * embedded devices.
+ */
+void *pvPortMalloc( size_t xWantedSize )
```
[the remark this answers is not in the document]
Comment (mmuesly) [reply tools/cbmc/include/cbmc.h]: I think, this discussion should be part of #774.

File: tools/cbmc/proofs/Queue/QueueGenericCreate/Configurations.json (hunk 3 above)
Comment (mmuesly) [reply tools/cbmc/proofs/Queue/QueueGenericCreate/Configurations.json]: The code is written in a way, that on boot the device uses pvPortMalloc to allocate a couple of data structures for reuse which are never freed while the device is only. Therefore the memory safety flags is not delivering very meaningful results as these data structures are supposed to be not freed.

File: tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c (hunk 4 above)
Comment (mmuesly) [reply tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c]: This is not yet specified. The function is supposed to be called with uxQueueLength > 0.

File: tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c (hunk 6 above)
Comment (mmuesly) [reply tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c]: portMAX_DELAY is the maximum value a UBaseType can get for a specific system. So this bound is resulting from the specs.

File: tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c (hunk 4 above)
Comment (danielsn) [reply tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c]: Is that documented in the mainline code, ideally as an assertion?

File: tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c (hunk 6 above)
Comment (danielsn) [reply tools/cbmc/proofs/Queue/QueueGenericCreate/QueueGenericCreate_harness.c]: That's a really odd name for the constant.  Is it possible to make a constant with a better name?

**Comment by markrtuttle** [comment]
Closed in favor of https://github.com/aws/amazon-freertos/pull/1324.
